controllers/space/space_controller.go

```diff
@@ -47,10 +47,10 @@ type Reconciler struct {
 func (r *Reconciler) SetupWithManager(mgr ctrl.Manager, memberClusters map[string]cluster.Cluster) error {
 	b := ctrl.NewControllerManagedBy(mgr).
 		// watch Spaces in the host cluster
-		For(&toolchainv1alpha1.Space{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		For(&toolchainv1alpha1.Space{}, builder.WithPredicates(predicate.GenerationChangedPredicate{}, predicate.AnnotationChangedPredicate{})).
```

Comment:
are you sure that this performs the logical `OR` operation on the predicates? I mean that the reconciliation is triggered either when the generation number is changed or when there is any changed in the annotations?

Reply:
Taking a quick glance at the controller-runtime code, this looks more like it does the logical `AND` operation over the set of predicates - it filters out all events that don't pass at least one of the provided predicates. In other words, the reconciliation is triggered only when the generation number is changed and at the same time when there is some change in the annotations. https://github.com/kubernetes-sigs/controller-runtime/blob/6ffcf19a340d9e2ee7b7738114c4d256f67d5337/pkg/source/internal/eventsource.go#L87-L91

That being said, we will need to create our own predicate for this - something similar to what we have for UserSignups https://github.com/codeready-toolchain/host-operator/blob/master/controllers/usersignup/predicate.go

Reply:
or there is even better solution - there is `predicate.Or` function
```suggestion
		For(&toolchainv1alpha1.Space{}, builder.WithPredicates(predicate.Or(predicate.GenerationChangedPredicate{}, predicate.AnnotationChangedPredicate{}))).
```

Reply:
I tested now with `OR` and it passed:

```
For(&toolchainv1alpha1.Space{}, builder.WithPredicates(
			predicate.Or(
				predicate.GenerationChangedPredicate{},
				predicate.AnnotationChangedPredicate{},
			))).
```

Reply:
I should have checked it... I just assumed it's `OR` by default :grimacing: